### PR TITLE
Add initial GSI-based soil analysis capability

### LINKED ISSUE #1479: Scripting for soil moisture and temperature analysis
**Description**

GSI issue  #508 (Add EnKF updating of soil temperature and soil moisture to the global EnKF) and the associated PR (#557) will add the capability to use the GSI EnKF to perform a soil moisture and/or temperature analysis. Implementing this at EMC will require changes to the global_workflow (including staging of obs, updates to anavinfo and convinfo, code to add increments to the model restarts ).

Note: addition of the soil increments to the model state is done outside of the GSI, and must be tailored to the land model (i.e, Noah, Noah-MP). Land increments have historically been added using global_cycle, however this code cannot handle fractional grids, so it may be better to add the increments using a purpose-built code base. Current code at PSL (not in global_cycle) for doing this for snow depth cannot interpolate between grids.

**Requirements**
Ability to run GSI (enkf only? or hybrid?) cycling experiments assimilating T2m and q2m to update model soil moisture and soil temperature.

**Acceptance Criteria (Definition of Done)**
Ability to run experiment described above without changing output from current experiments.

### PR DESCRIPTION
**Pull request by ClaraDraper-NOAA**
# Description

First set of changes for adding the new soil analysis, from the assimilation of screen-level T and q. The changes here enable the screen-level observations to be assimilated in the Hybrid (Var and EnKF) update, and the soil temperature and soil moisture updates to be made in the EnKF only.  The functionality is turned on by setting GSI_SOILANAL to YES in config.base.

Resolves #1479 

# Type of change

- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?

# Checklist
- [ ] Any dependent changes have been merged and published
- [ x] My code follows the style guidelines of this project
- [x ] I have performed a self-review of my own code
- [ x] I have commented my code, particularly in hard-to-understand areas
- [ x] My changes generate no new warnings
- [ x] New and existing tests pass with my changes
- [x ] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

**Comment by ClaraDraper-NOAA** [comment]
@aerorahul @CatherineThomas-NOAA  @KateFriedman-NOAA 

I've created this as a draft for now, as it includes my junky code to get around the UFS_UTILS / Noah-MP problem. Once global_workflow can handle Noah-MP, I'll update my code and submit the PR. 

Note: this PR is also dependent on an update to the GSI (PR under review), and updating the UFS_UTILS hash (but updating UFS_UTILS for Noah-MP will take care of the latter, as my hash is before that one).

One question:  The soil analysis requires changes to convinfo and anavinfo. At the moment those are hard-linked to my directory on hera. How would you like me to handle that?

**Comment by WalterKolczynski-NOAA** [comment]
@ClaraDraper-NOAA Are these new files? If so they should be added to `parm/`. I don't think they fit in any existing directory, so let's make a new one tentatively called `parm/soil_da` (will discuss name with Rahul and Kate). This will make the lines:

```
export CONVINFO="${HOMEgfs}/parm/soil_da/global_convinfos.w2mobs.txt"
export ANAVINFO="${HOMEgfs}/parm/soil_da/global_anavinfo.2mobs_soilupd.l127.txt"
```

If you want to save time later, go ahead and address the shellcheck warnings while waiting for the other PRs. Please reach out if you need help.

**Comment by ClaraDraper-NOAA** [comment]
> @ClaraDraper-NOAA Are these new files? If so they should be added to `parm/`. I don't think they fit in any existing directory, so let's make a new one tentatively called `parm/soil_da` (will discuss name with Rahul and Kate). This will make the lines:
> 
> ```
> export CONVINFO="${HOMEgfs}/parm/soil_da/global_convinfos.w2mobs.txt"
> export ANAVINFO="${HOMEgfs}/parm/soil_da/global_anavinfo.2mobs_soilupd.l127.txt"
> ```
> 
> If you want to save time later, go ahead and address the shellcheck warnings while waiting for the other PRs. Please reach out if you need help.

Thanks @WalterKolczynski-NOAA . These are updates to fix files (global_convinfo.txt and global_anavinfo.txt, in $FIXgsi). Perhaps it'd be better if I just noted in the PR that those files need to be changed if running the soil analysis?

I'll sort out the spelling mistakes soon.

**Comment by KateFriedman-NOAA** [comment]
> These are updates to fix files (global_convinfo.txt and global_anavinfo.txt, in $FIXgsi).

@ClaraDraper-NOAA FYI, we will soon have an updated version of the GSI fix files that includes the updates from the operational updates for NOAA21. I am tagging @RussTreadon-NOAA and @ADCollard just in case this conflicts with that update or needs to be handled carefully. Also want to make sure we don't blow away your update when the ops updates come in. See issue #1919 for work that is on hold but planned to do soon.

**Comment by ADCollard** [comment]
> > These are updates to fix files (global_convinfo.txt and global_anavinfo.txt, in $FIXgsi).
> 
> @ClaraDraper-NOAA FYI, we will soon have an updated version of the GSI fix files that includes the updates from the operational updates for NOAA21. I am tagging @RussTreadon-NOAA and @ADCollard just in case this conflicts with that update or needs to be handled carefully. Also want to make sure we don't blow away your update when the ops updates come in. See issue #1919 for work that is on hold but planned to do soon.

@KateFriedman-NOAA @ClaraDraper-NOAA I don't see there being any significant conflicts here (and will be careful when adding in the NOAA-21 updates).   We are still awaiting a correct version of the CRTM on spack-stack, but hopefully that will be fixed soon (see https://github.com/JCSDA/spack/pull/399)

**Comment by RussTreadon-NOAA** [comment]
@ClaraDraper-NOAA , where may I find the GSI-fix files with your changes?  ASCII GSI fix files are maintained in N[OAA-EMC/GSI-fix](https://github.com/NOAA-EMC/GSI-fix).  Given the location of your changes to `global_convinfo.txt` and `global_anavinfo.l127.txt`, we can commit your changes to a GSI-fix branch.

GSI-fix issue [#11](https://github.com/NOAA-EMC/GSI-fix/issues/11) documents the addition of operational GFS v16.3.12 ASCII GSI fix files to [GSI-fix develop](https://github.com/NOAA-EMC/GSI-fix/tree/develop).  Would you please open a [new GSI-fix issue](https://github.com/NOAA-EMC/GSI-fix/issues) to document the addition of the ASCII fix file changes required for the initial GSI-based soil DA capability.

**Comment by ClaraDraper-NOAA** [comment]
@RussTreadon-NOAA The new fix files are on hera, in this directory: 

/scratch2/BMC/gsienkf/Clara.Draper/2mDAfiles_gworkflow/

The changes to convinfo will add the assimilation of screen-level observations (although they will then all be rejected, until a change is made in the obs prep), and the changes to the anavinfo will add the soil states to the control vector. 

I'll open an issue in the fix repo as soon as I get a chance (at AMS this week).

**Comment by RussTreadon-NOAA** [comment]
Thank you @ClaraDraper-NOAA .  The changes to anavinfo and convinfo look straightforward.   Opening a [GSI-fix](https://github.com/NOAA-EMC/GSI-fix/issues) issue after AMS works.

**Comment by KateFriedman-NOAA** [comment]
Thanks @RussTreadon-NOAA ! Let me know when there is a new version of GSI-fix that is ready to be added to the fix sets. It would be great if this update could be combined with the GFSv16.3.11[12] fix updates and we can have a single new GSI-fix timestamp in fix. No worries if not though! :)

**Comment by RussTreadon-NOAA** [comment]
@KateFriedman-NOAA , I'll let @ClaraDraper-NOAA and @ADCollard coordinate with you on upcoming changes to GSI fix files. 
 We have changes to GSI ASCII fix files and GSI binary files.   Updates to [GSI-fix](https://github.com/NOAA-EMC/gsi-fix) will track changes to ASCII GSI-fix.  GSI binaries will pass through the [Fix File Update issue](https://github.com/NOAA-EMC/global-workflow/issues/new?assignees=KateFriedman-NOAA%2CWalterKolczynski-NOAA&labels=Fix+Files%2Ctriage&projects=&template=fix_file.yml) in g-w.

**Comment by ClaraDraper-NOAA** [comment]
GSIfix issue is here: [12](https://github.com/NOAA-EMC/GSI-fix/issues/12).

**Comment by ClaraDraper-NOAA** [comment]
Following discussion with @CatherineThomas-NOAA and discussion in GSI-fix issue [12](https://github.com/NOAA-EMC/GSI-fix/issues/12) I've submitted PR [14](https://github.com/NOAA-EMC/GSI-fix/pull/14) with the new convinfo and anavinfo files required for this PR included as alternatives for those experimenting with the soil analysis.

**Comment by ClaraDraper-NOAA** [comment]
> @ClaraDraper-NOAA Are these new files? If so they should be added to `parm/`. I don't think they fit in any existing directory, so let's make a new one tentatively called `parm/soil_da` (will discuss name with Rahul and Kate). This will make the lines:
> 
> ```
> export CONVINFO="${HOMEgfs}/parm/soil_da/global_convinfos.w2mobs.txt"
> export ANAVINFO="${HOMEgfs}/parm/soil_da/global_anavinfo.2mobs_soilupd.l127.txt"
> ```
> 
> If you want to save time later, go ahead and address the shellcheck warnings while waiting for the other PRs. Please reach out if you need help.

@WalterKolczynski-NOAA I'm now getting spellcheck warnings from the discover commits that I just merged in. Do you want me to fix these?

**Comment by ClaraDraper-NOAA** [comment]
@KateFriedman-NOAA @WalterKolczynski-NOAA I think this PR is now ready for review. I'll start testing that it runs with and without the soil analysis turned on. Please let me know if there's anything else I should do before submitting it.

@CoryMartin-NOAA 

 The necessary GSI PR has been merged (but the hash in global_workflow has not been updated), and the UFS_UTILS changes has been merged (and hash updated). The outstanding dependency is the GSI-fix PR referenced above.

**Comment by RussTreadon-NOAA** [comment]
Fix files 
```
global_anavinfo_soilanal.l127.txt
global_convinfo_2mObs.txt
```
have been added to [GSI-fix](https://github.com/NOAA-EMC/GSI-fix) `develop`.  Done at [`298bdc0`](https://github.com/NOAA-EMC/GSI-fix/commit/298bdc0f56692ad25141a946177e5d88884e367a)

File: scripts/exgdas_enkf_sfc.sh
Comment (WalterKolczynski-NOAA) [inline scripts/exgdas_enkf_sfc.sh:160]:
```suggestion
                ${NLN} "${COM_ATMOS_ANALYSIS_MEM}/${APREFIX_ENS}sfci006.nc" \
```

File: scripts/exgdas_enkf_update.sh
Comment (WalterKolczynski-NOAA) [inline scripts/exgdas_enkf_update.sh:209]: These files really should have three-digit forecast hours like everywhere else, but since this matches the existing sfg files, it is fine for now.

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: scripts/exgdas_enkf_sfc.sh
Comment (CoryMartin-NOAA) [reply scripts/exgdas_enkf_sfc.sh:160]: the consistent change would be this becomes `sfcinc.nc`, I suggest if we want `sfci006` that we do that in a subsequent PR and do it system-wide.

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Conditionally approved pending successful completion of CI tests.

**Comment by WalterKolczynski-NOAA** [comment]
This branch needs to pull in changes from develop.

**Comment by WalterKolczynski-NOAA** [comment]
@ClaraDraper-NOAA Please merge in current development and resolve any conflicts. Then we should be able to run CI tests on it.

**Comment by ClaraDraper-NOAA** [comment]
Thanks @WalterKolczynski-NOAA I'm on leave today, but will get started tomorrow. There's also a small bug I need to fix, so I'd like to run my own test to make sure everything is OK too.

**Comment by WalterKolczynski-NOAA** [comment]
@ClaraDraper-NOAA Sorry, one more time. We merged a CI change. There shouldn't be any conflicts.

**Comment by ClaraDraper-NOAA** [comment]
> @ClaraDraper-NOAA Sorry, one more time. We merged a CI change. There shouldn't be any conflicts.

Walter - thanks for letting me know (still compiling, it's sloooow today). I'll update now. I'm ignoring the spell check warnings, as they don't relate to my changes.

**Comment by ClaraDraper-NOAA** [comment]
@WalterKolczynski-NOAA @RussTreadon-NOAA With GSI_SOILANAL set to YES (not the default), we need to update the gsi fix directory from 20230911 to 20240208,  to pick up the convinfo and anavinfo we recently added. It looks like GSI_BINARY_SOURCE_DIR is set in the gsi*lua files. How would you like me to handle this? For my tests (on hera), I'm just manually updating the link.

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA @RussTreadon-NOAA With GSI_SOILANAL set to YES (not the default), we need to update the gsi fix directory from 20230911 to 20240208, to pick up the convinfo and anavinfo we recently added. It looks like GSI_BINARY_SOURCE_DIR is set in the gsi*lua files. How would you like me to handle this? For my tests (on hera), I'm just manually updating the link.

The `versions/fix.ver` file needs to be updated to use the new version.

**Comment by WalterKolczynski-NOAA** [comment]
Killed CI until the fix gets updated.

File: parm/config/gfs/config.base.emc.dyn
Comment (aerorahul) [inline parm/config/gfs/config.base.emc.dyn]: Can we make this a templated variable and set it to `NO` as the default?
That way, we can set this to `YES` in one of the existing tests and ensure this capability is tested regularly.
The test this could piggy back off on is [C96C48_hybatmDA.yaml](https://github.com/NOAA-EMC/global-workflow/blob/develop/ci/cases/pr/C96C48_hybatmDA.yaml)

Comment (ClaraDraper-NOAA) [reply parm/config/gfs/config.base.emc.dyn]: Rahul - I'm not familiar with the templating, but followed through the example for DOIAU. Is this all I need to do? 

In config.base.emc.dyn change the line highlighted above to:  

export GSI_SOILANAL=@GSI_SOILANAL@

Then in parm/config/gfs/yaml/defaults.yaml  I set the default to NO. 

For the test, should I add GSI_SOILANAL=YES to [C96C48_hybatmDA.yaml](https://github.com/NOAA-EMC/global-workflow/blob/develop/ci/cases/pr/C96C48_hybatmDA.yaml), or create a new test based on that one?

**Comment by ClaraDraper-NOAA** [comment]
I've updated the fix version, templated GSI_SOILANAL, and checked that the land DA is working as expected. It's all ready to go from my end.

I haven't added/updated the test. However, this PR is only the first step for the land DA (apply increments to the ensemble members). The next PR (which will take me 1-2 months) will then also update the deterministic member. We could wait until then to add  this to a test?

**Comment by aerorahul** [comment]
Apologies @ClaraDraper-NOAA 
This slipped past me.
1. I think we can keep GSI_SOILANAL="NO" till such time you and others are comfortable to make this `YES` and then we will keep it permanently `YES` unless there is a desire to turn it off.
2. Please update your branch by resolving the conflicts and we can run this through the CI and prepare to merge.

**Review by aerorahul — APPROVED** [review]
looks good.

**Comment by ClaraDraper-NOAA** [comment]
Thanks @aerorahul . I'll get to it tomorrow.

**Comment by ClaraDraper-NOAA** [comment]
@aerorahul I've solved the merge conflicts. We should be good to start the CI tests.

**Pull request merged**
